### ISSUE TITLE
No documentation for headless Firefox

### ISSUE THREAD

**Issue by ThomasG77** [webdriverio/webdriverio#2299]
## The problem: no documentation for headless Firefox

I've found documentation for Chrome headless mode at https://github.com/webdriverio/webdriverio/blob/master/docs/guide/testrunner/configurationfile.md but there are none for Firefox headless

Firefox `capabilities` section in the same `configurationfile.md` file should mention how to do it

Hence,

```
        // maxInstances can get overwritten per capability. So if you have an in house Selenium
        // grid with only 5 firefox instance available you can make sure that not more than
        // 5 instance gets started at a time.
        maxInstances: 5,
        browserName: 'firefox',
        specs: [
            'test/ffOnly/*'
        ]
```

should now be :

```
        // maxInstances can get overwritten per capability. So if you have an in house Selenium
        // grid with only 5 firefox instance available you can make sure that not more than
        // 5 instance gets started at a time.
        maxInstances: 5,
        browserName: 'firefox',
        "moz:firefoxOptions": {
          // flag to activate Firefox headless mode (see https://github.com/mozilla/geckodriver/blob/master/README.md#firefox-capabilities for more details about moz:firefoxOptions)
          // args: ['-headless']
        }
        specs: [
            'test/ffOnly/*'
        ]
```

Can you confirm it works on your side? If ok, I can make a PR if you agree with my suggestion.

## Environment

* Webdriverio version 4.8.0
* Node.js version v8.4.0
* wdio testrunner configured with `services: ['selenium-standalone'],`

**Comment by christian-bromann** [comment]
https://github.com/webdriverio/webdriverio/commit/5aa704acf0946a51203204c5324e1429ae7dd5ac

**Comment by ThomasG77** [comment]
Thanks for integrating my suggestion

**Comment by SavePointSam** [comment]
For anyone ending up here in the future having difficulty getting headless to work. Note that `-headless` wasn't implemented until Firefox 55 for Linux and Firefox 56 on MacOS/Windows.

The current Firefox for CentOS 7 via yum is 52. I was stuck on that for a couple hours.

**Comment by iongion** [comment]
Still having this issue

*Firefox* `63.0.3`
*wdio* `v4.14.1`

Verbose log of wdio:

```
...
[10:22:01]  DATA {
"desiredCapabilities" : {
 "javascriptEnabled":true,
 "locationContextEnabled":true,
 "handlesAlerts":true,
 "rotatable":true,
 "maxInstances":1,
 "browserName":"firefox",
 "moz:firefoxOptions":{
   "args":["-headless"]
 },
 "loggingPrefs":{
   "browser":"ALL",
   "driver":"ALL"
  },
 "requestOrigins":{
    "url":"http://webdriver.io",
    "version":"4.14.1",
    "name":"webdriverio"
  }
}}
[10:22:01]  INFO	SET SESSION ID 5843972a839b897a06180ce73f344eca
[10:22:01]  RESULT		{
   "acceptInsecureCerts":false,
   "acceptSslCerts":false,
   "applicationCacheEnabled":false,
   "browserConnectionEnabled":false,
   "browserName":"chrome",
   "chrome":{
      "chromedriverVersion":"2.44.609551 (5d576e9a44fe4c5b6a07e568f1ebc753f1214634)",
      "userDataDir":"/tmp/.org.chromium.Chromium.sZ5ksp"
   },
   "cssSelectorsEnabled":true,
   "databaseEnabled":false,
   "goog:chromeOptions":{
      "debuggerAddress":"localhost:38103"
   },
   "handlesAlerts":true,
   "hasTouchScreen":false,
   "javascriptEnabled":true,
   "locationContextEnabled":true,
   "mobileEmulationEnabled":false,
   "nativeEvents":true,
   "networkConnectionEnabled":false,
   "pageLoadStrategy":"normal",
   "platform":"Linux",
   "rotatable":false,
   "setWindowRect":true,
   "takesHeapSnapshot":true,
   "takesScreenshot":true,
   "unexpectedAlertBehaviour":"ignore",
   "version":"71.0.3578.80",
   "webStorageEnabled":true
}
...
```

Tests run, but I still *SEE* the Firefox window appearing on my screen.

The `wdio.conf.js`

```javascript
'use strict';

const baseUrl = 'http://localhost:8081';

exports.config = {

    port: '9515',
    path: '/',

    chromeDriverLogs: './logs',

    //
    // ==================
    // Specify Test Files
    // ==================
    // Define which test specs should run. The pattern is relative to the directory
    // from which `wdio` was called. Notice that, if you are calling `wdio` from an
    // NPM script (see https://docs.npmjs.com/cli/run-script) then the current working
    // directory is where your package.json resides, so `wdio` will be called from there.
    //
    specs: [
        './test/spec/**/*.spec.js',
    ],
    // Patterns to exclude.
    exclude: [
        // 'path/to/excluded/files'
    ],
    //
    // ============
    // Capabilities
    // ============
    // Define your capabilities here. WebdriverIO can run multiple capabilities at the same
    // time. Depending on the number of capabilities, WebdriverIO launches several test
    // sessions. Within your capabilities you can overwrite the spec and exclude options in
    // order to group specific specs to a specific capability.
    //
    // First, you can define how many instances should be started at the same time. Let's
    // say you have 3 different capabilities (Chrome, Firefox, and Safari) and you have
    // set maxInstances to 1; wdio will spawn 3 processes. Therefore, if you have 10 spec
    // files and you set maxInstances to 10, all spec files will get tested at the same time
    // and 30 processes will get spawned. The property handles how many capabilities
    // from the same test should run tests.
    //
    maxInstances: 1,
    //
    // If you have trouble getting all important capabilities together, check out the
    // Sauce Labs platform configurator - a great tool to configure your capabilities:
    // https://docs.saucelabs.com/reference/platforms-configurator
    //
    capabilities: [
      // maxInstances can get overwritten per capability. So if you have an in-house Selenium
      // grid with only 5 firefox instances available you can make sure that not more than
      // 5 instances get started at a time.
      /*
      {
        maxInstances: 1,
        //
        browserName: 'chrome',
        chromeOptions: {
          args: [
            '--headless',
            '--disable-gpu',
            '--no-sandbox'
          ],
        }
      },
      */
      {
        maxInstances: 1,
        //
        browserName: 'firefox',
        'moz:firefoxOptions': {
          args: ['-headless'],
        },
      },
    ],
    //
    // ===================
    // Test Configurations
    // ===================
    // Define all options that are relevant for the WebdriverIO instance here
    //
    // By default WebdriverIO commands are executed in a synchronous way using
    // the wdio-sync package. If you still want to run your tests in an async way
    // e.g. using promises you can set the sync option to false.
    sync: true,
    //
    // Level of logging verbosity: silent | verbose | command | data | result | error
    logLevel: 'verbose',
    //
    // Enables colors for log output.
    coloredLogs: true,
    //
    // Warns when a deprecated command is used
    deprecationWarnings: true,
    //
    // If you only want to run your tests until a specific amount of tests have failed use
    // bail (default is 0 - don't bail, run all tests).
    bail: true,
    //
    // Saves a screenshot to a given path if a command fails.
    screenshotPath: './errorShots/',
    //
    // Set a base URL in order to shorten url command calls. If your `url` parameter starts
    // with `/`, the base url gets prepended, not including the path portion of your baseUrl.
    // If your `url` parameter starts without a scheme or `/` (like `some/path`), the base url
    // gets prepended directly.
    baseUrl,
    //
    // Default timeout for all waitFor* commands.
    waitforTimeout: 30000,
    //
    // Default timeout in milliseconds for request
    // if Selenium Grid doesn't send response
    connectionRetryTimeout: 90000,
    //
    // Default request retries count
    connectionRetryCount: 3,
    //
    // Initialize the browser instance with a WebdriverIO plugin. The object should have the
    // plugin name as key and the desired plugin options as properties. Make sure you have
    // the plugin installed before running any tests. The following plugins are currently
    // available:
    // WebdriverCSS: https://github.com/webdriverio/webdrivercss
    // WebdriverRTC: https://github.com/webdriverio/webdriverrtc
    // Browserevent: https://github.com/webdriverio/browserevent
    // plugins: {
    //     webdrivercss: {
    //         screenshotRoot: 'my-shots',
    //         failedComparisonsRoot: 'diffs',
    //         misMatchTolerance: 0.05,
    //         screenWidth: [320,480,640,1024]
    //     },
    //     webdriverrtc: {},
    //     browserevent: {}
    // },
    //
    // Test runner services
    // Services take over a specific job you don't want to take care of. They enhance
    // your test setup with almost no effort. Unlike plugins, they don't add new
    // commands. Instead, they hook themselves up into the test process.
    services: ['chromedriver'],
    //
    // Framework you want to run your specs with.
    // The following are supported: Mocha, Jasmine, and Cucumber
    // see also: http://webdriver.io/guide/testrunner/frameworks.html
    //
    // Make sure you have the wdio adapter package for the specific framework installed
    // before running any tests.
    framework: 'mocha',
    //
    // Test reporter for stdout.
    // The only one supported by default is 'dot'
    // see also: http://webdriver.io/guide/reporters/dot.html
    reporters: ['spec'],

    //
    // Options to be passed to Mocha.
    // See the full list at http://mochajs.org/
    mochaOpts: {
        ui: 'bdd'
    },
    //
    // =====
    // Hooks
    // =====
    // WebdriverIO provides several hooks you can use to interfere with the test process in order to enhance
    // it and to build services around it. You can either apply a single function or an array of
    // methods to it. If one of them returns with a promise, WebdriverIO will wait until that promise got
    // resolved to continue.
    /**
     * Gets executed once before all workers get launched.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     */
    // onPrepare: function (config, capabilities) {
    // },
    /**
     * Gets executed just before initialising the webdriver session and test framework. It allows you
     * to manipulate configurations depending on the capability or spec.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that are to be run
     */
    // beforeSession: function (config, capabilities, specs) {
    // },
    /**
     * Gets executed before test execution begins. At this point you can access to all global
     * variables like `browser`. It is the perfect place to define custom commands.
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that are to be run
     */
    before: function (capabilities, specs) {
      var chai = require('chai');
      chai.use(require('chai-as-promised'));
      global.expect = chai.expect;
      global.assert = chai.assert;
      global.project = require('./package.json');
      global.project.baseUrl = baseUrl;
      chai.should();
    },
    /**
     * Runs before a WebdriverIO command gets executed.
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     */
    // beforeCommand: function (commandName, args) {
    // },

    /**
     * Hook that gets executed before the suite starts
     * @param {Object} suite suite details
     */
    // beforeSuite: function (suite) {
    // },
    /**
     * Function to be executed before a test (in Mocha/Jasmine) or a step (in Cucumber) starts.
     * @param {Object} test test details
     */
    // beforeTest: function (test) {
    // },
    /**
     * Hook that gets executed _before_ a hook within the suite starts (e.g. runs before calling
     * beforeEach in Mocha)
     */
    // beforeHook: function () {
    // },
    /**
     * Hook that gets executed _after_ a hook within the suite ends (e.g. runs after calling
     * afterEach in Mocha)
     */
    // afterHook: function () {
    // },
    /**
     * Function to be executed after a test (in Mocha/Jasmine) or a step (in Cucumber) ends.
     * @param {Object} test test details
     */
    // afterTest: function (test) {
    // },
    /**
     * Hook that gets executed after the suite has ended
     * @param {Object} suite suite details
     */
    // afterSuite: function (suite) {
    // },

    /**
     * Runs after a WebdriverIO command gets executed
     * @param {String} commandName hook command name
     * @param {Array} args arguments that command would receive
     * @param {Number} result 0 - command success, 1 - command error
     * @param {Object} error error object if any
     */
    // afterCommand: function (commandName, args, result, error) {
    // },
    /**
     * Gets executed after all tests are done. You still have access to all global variables from
     * the test.
     * @param {Number} result 0 - test pass, 1 - test fail
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // after: function (result, capabilities, specs) {
    // },
    /**
     * Gets executed right after terminating the webdriver session.
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     * @param {Array.<String>} specs List of spec file paths that ran
     */
    // afterSession: function (config, capabilities, specs) {
    // },
    /**
     * Gets executed after all workers got shut down and the process is about to exit.
     * @param {Object} exitCode 0 - success, 1 - fail
     * @param {Object} config wdio configuration object
     * @param {Array.<Object>} capabilities list of capabilities details
     */
    // onComplete: function(exitCode, config, capabilities) {
    // }
}

```

The deps:

```
    "chai": "^4.2.0",
    "chai-as-promised": "^7.1.1",
    "chai-string": "^1.5.0",
    "chromedriver": "^2.44.1",
    "mocha": "^5.2.0",
    "wdio-chromedriver-service": "^0.1.5",
    "wdio-firefox-profile-service": "^0.1.3",
    "wdio-mocha-framework": "^0.6.4",
    "wdio-spec-reporter": "^0.1.5",
    "webdriverio": "^4.14.1"

```

What am I missing ?

**Comment by Christopher-Hayes** [comment]
Hey, just as a heads up for anyone else looking for a way to test on Firefox, there is an addon/service for WebDriver.io this random dude made that is supposed to work like `wdio-chromedriver-service`. For one thing, on Firefox I think you want `geckodriver`, Mozilla has an official repo for Selenium, but I couldn't find anything that easily integrates this into WebDriver.io. Until I came across this [https://github.com/RafalSkorka/wdio-geckodriver-service](https://github.com/RafalSkorka/wdio-geckodriver-service), which acts as a wrapper around Mozilla's `webdriver`. It doesn't have any stars and looks pretty new (4 months), so you've been warned. But, it seems to be working for me, the repo looks like it was largely based off of `wdio-chromedriver-service` and it acts a lot like the chrome service. I added the service: `services: ['geckodriver']` and it seems to work fine with very minimal setup.
